sql, sql/parser, sql/sem/tree: simplify the handling of names

knz opened the pull request:
Forked off #21753 to simplify the review.

Prior to this patch, a single non-terminal `qualified_name` in the
grammar was handling all the other grammar needs for naming things
with a path that optionally includes a star (x.y.z), `any_name` for a
path that cannot include a star, and `name` and `unrestricted_name`
were sprinkled around the other rules for things without a path,
without apparent structure.

In a future patch we want to properly discriminate between names that
designate a database object (with a possible db and schema prefix)
from those names that designate other things, so as to only apply
the schema vs catalog normalization rules on the former category.

As a prerequisite to this change, we want the grammar to properly
reveal which name belongs to which category.

This patch does just this, by using different non-terminals in the
grammar for different kinds of names:

- `table_name` for a potentially-qualified table name with no star
- `table_pattern` for a potentially-qualified table name or glob
- `column_name` for a single unqualified column name
- `column_path` for a potentially-qualified column name with no star
- `column_path_with_star` for a potentially-qualified column name or glob
- `var_name` for a variable name potentially containing periods

etc.

In addition, it avoids a parsing recursion for `table_name`,
`column_path` variants and `table_pattern` and reveals the various
fixed formats that the various names can have. In various places,
starting with function names, a star is now rejected during parsing
as a syntax error instead of during name resolution (= less work).

While authoring this patch, two side effects were obtained:

- the handling of INSERT, UPDATE and UPSERT was simplified, by
  avoiding paying the cost of name normalization for the target column
  list.  Since CockroachDB currently only supports simple names in
  this position, the grammar is simplified accordingly.
- a bug in the grammar accepted for COPY was fixed: the grammar
  previously recognized the syntax "`COPY foo() FROM STDIN`", whereas
  an empty column list between the parentheses is incorrect according
  to pg's grammar.

Release note: None

spencerkimball committed:
kv: refresh spans on SERIALIZABLE txns to avoid full restarts

The concept is straightforward: store spans encountered during
the course of executing a transaction, up to a maximum limit of bytes.
Upon encountering a transaction retry error with `reason=RETRY_SERIALIZABLE`,
attempt to "refresh" all of the spans in order to avoid a full txn
restart. For the case of write-only SERIALIZABLE transactions, we're now
able to set the original timestamp equal to the commit timestamp before
committing, entirely circumventing the possibility of a serializable
restart. Otherwise, if all spans can be updated, the original timestamp
is moved to the commit timestamp (without a transaction restart), and the
`EndTransaction` batch is re-submitted.

`Refresh(Range)` is a new KV RPC which re-reads a span, updating the timestamp
cache and checking whether any more recent writes occurred to the span.
It uses a time-bounded iterator for efficiency. Any writes or intents
which have appeared between the transaction's original timestamp and its
commit timestamp result in an error. Note that `Refresh(Range)` has a bool
indicating whether it should update the read or write timestamp cache.
Only `DeleteRange` creates update spans which need to update the write
timestamp cache.

This change should allow long-lived SERIALIZABLE transactions in the same
style as SNAPSHOT, and should decrease the occurence of transaction retry
errors which currently must be handled by client code.

See #21078

Release note (sql): significantly reduce the likelihood of serializable
restarts seen by clients due to concurrent workloads.

knz committed:
sql, sql/parser, sql/sem/tree: simplify the handling of names

Prior to this patch, a single non-terminal `qualified_name` in the
grammar was handling all the other grammar needs for naming things
with a path that optionally includes a star (x.y.z), `any_name` for a
path that cannot include a star, and `name` and `unrestricted_name`
were sprinkled around the other rules for things without a path,
without apparent structure.

In a future patch we want to properly discriminate between names that
designate a database object (with a possible db and schema prefix)
from those names that designate other things, so as to only apply
the schema vs catalog normalization rules on the former category.

As a prerequisite to this change, we want the grammar to properly
reveal which name belongs to which category.

This patch does just this, by using different non-terminals in the
grammar for different kinds of names:

- `table_name` for a potentially-qualified table name with no star
- `table_pattern` for a potentially-qualified table name or glob
- `column_name` for a single unqualified column name
- `column_path` for a potentially-qualified column name with no star
- `column_path_with_star` for a potentially-qualified column name or glob
- `var_name` for a variable name potentially containing periods

etc.

In addition, it avoids a parsing recursion for `table_name`,
`column_path` variants and `table_pattern` and reveals the various
fixed formats that the various names can have. In various places,
starting with function names, a star is now rejected during parsing
as a syntax error instead of during name resolution (= less work).

While authoring this patch, two side effects were obtained:

- the handling of INSERT, UPDATE and UPSERT was simplified, by
  avoiding paying the cost of name normalization for the target column
  list.  Since CockroachDB currently only supports simple names in
  this position, the grammar is simplified accordingly.
- a bug in the grammar accepted for COPY was fixed: the grammar
  previously recognized the syntax "`COPY foo() FROM STDIN`", whereas
  an empty column list between the parentheses is incorrect according
  to pg's grammar.

Release note: None

knz committed:
sql/sem/tree: make IMPORT use normalizable table references

... for consistency with every other use of `table_name`.

Release note: None

knz committed:
sql, sql/parser, sql/sem/tree: sanitize/simplify the handling of names

Prior to this patch, the type `UnresolvedName` was an array of things
called "name parts", and could include stars, array indirections or
subscripts. The corresponding logic to normalize unresolved names into
table names, column names or other things was complicated.

In the light of a desired upcoming change to support both catalog and
schema specifiations in names, this code was welcoming a
simplification.

The simplification came from the realization that the syntactic
structure for a table or column path is unambiguous and the "item of
interest" is always in the last position. My previous (mistaken)
understanding was that a column name could be followed by other name
parts in the grammar; however, this is not valid SQL.

- to access an array column, the grammar rules to parse `a.b.c[d]`
  parse first `a.b.c` as a column item, then `<expr> [d]` as an array
  subscript expression. The `[d]` part is not part of the grammar for
  names, never has.

- to access a field in a column of a compound type (these are not
  supported by CockroachDB yet, but the grammar mirrors pg's which
  does support them), pg's SQL dialect has it that one *must* write
  `(a.b.c).field` with grouping parentheses; a direct path of
  `a.b.c.field` is not valid. (More precisely, `a.b.c.field`
  *always* refers to the column `field` of table `c` in schema `b` in
  catalog `a`, and the query would error out if there is no such
  table/column.)

  In short, an unresolved column name again always has a period-separated
  path with the column name in last position.

The predecessor to this patch already had simplified the grammar rules
to account for the known maximum length of a qualified name. This
patch now simplifies the AST and all uses to further capitalize on the
new understanding and simplify the normalization and pretty-printing
rules.

Release note: None

knz committed:
sql/sem/tree, sql: clarify TableName.SchemaName and TableName.CatalogName

Working slowly towards a better use of the catalog prefix in `TableName`,
this patch clarifies the struct:

- the "Prefix" field is renamed to "Catalog".
- the fields "OmitSchemaDuringFormatting" and
  "PrefixOriginallySpecified" are renamed to "ExplicitSchema" (meaning
  inverted) and "ExplicitCatalog", to clarify what they are really
  about.
- new helper methods are provided to construct instances of
  `TableName`. This sensibly shortens most points of use, and
  clarifies intent.

Release note: None

couchand committed:
ui: remove unused clusterviz files

Release note: None

couchand committed:
ui: remove prototype treemap from clusterviz

Release note: None

Commit:
server/ui: Compute LivenessStatus on Server

Moves the computation of node "liveness status" from the Admin UI to the
server.

As part of this, an additional liveness status has been added on the
server: "Decommissioning" has been split into "Decommissioning" and
"Decommissioned". The second status is given to decommissioned nodes
with an expired liveness record.

On the server, the "Decommissioned" state is treated equivalently to
Dead, as was always the case. However, the UI uses the "Decommissoned"
state to hide nodes from the UI. On the UI, "Decommissioning" is now
counted as a suspect node, although it does get custom tooltip text and
custom text on the single-node page.

+ Admin API "Liveness" method now returns a map from NodeID to
LivenessStatus. This utilizes the same code path that was previously
being used by the Store Pool to categorize store availability.
+ Admin UI now uses NodeLivenessStatus enumeration defined in
liveness.proto; previously, the Admin UI had defined its own
enumeration.

Fixes #18896
Fixes #17840

Release note: Fixed a bug where healthy nodes could appear as "Suspect"
in the Admin UI if the web browser's local clock was skewed.

jordanlewis committed:
storage: append-only data struct for MVCCScan

Previously, the C++ MVCCScan code wrote the data it found on disk into a
std::string, which copies and reallocates when it runs out of capacity.
This was inefficient, since we eventually just copy it into Go memory
anyway.

Now, MVCCScan writes to a new data structure called chunkedBuffer, which
maintains a vector of buffers that increase exponentially in size
starting at 16 bytes. When one of the buffers runs out of capacity, a
new one is created and appended to the vector.

Then, Go copies data out of this sequence of buffers into Go memory.

```
name                                                    old time/op    new time/op    delta
MVCCScan_RocksDB/rows=1/versions=1/valueSize=8-8          5.45µs ± 2%    5.33µs ± 2%   -2.23%  (p=0.000 n=10+9)
MVCCScan_RocksDB/rows=1/versions=1/valueSize=64-8         5.60µs ± 2%    5.64µs ± 2%     ~     (p=0.280 n=10+10)
MVCCScan_RocksDB/rows=1/versions=1/valueSize=512-8        6.48µs ± 2%    6.41µs ± 3%     ~     (p=0.271 n=10+10)
MVCCScan_RocksDB/rows=10/versions=1/valueSize=8-8         8.38µs ± 1%    8.27µs ± 2%   -1.34%  (p=0.003 n=10+10)
MVCCScan_RocksDB/rows=10/versions=1/valueSize=64-8        9.12µs ± 3%    8.94µs ± 1%   -2.01%  (p=0.001 n=9+10)
MVCCScan_RocksDB/rows=10/versions=1/valueSize=512-8       12.0µs ± 1%    11.7µs ± 1%   -2.21%  (p=0.000 n=10+9)
MVCCScan_RocksDB/rows=100/versions=1/valueSize=8-8        31.8µs ± 3%    28.9µs ± 1%   -9.14%  (p=0.000 n=9+10)
MVCCScan_RocksDB/rows=100/versions=1/valueSize=64-8       34.5µs ± 2%    31.8µs ± 1%   -7.66%  (p=0.000 n=10+10)
MVCCScan_RocksDB/rows=100/versions=1/valueSize=512-8      50.6µs ± 3%    45.8µs ± 1%   -9.54%  (p=0.000 n=10+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=8-8        251µs ± 1%     221µs ± 0%  -12.25%  (p=0.000 n=10+9)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=64-8       269µs ± 0%     237µs ± 1%  -11.95%  (p=0.000 n=9+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=512-8      425µs ± 1%     371µs ± 1%  -12.53%  (p=0.000 n=10+9)
MVCCScan_RocksDB/rows=10000/versions=1/valueSize=8-8      2.48ms ± 2%    2.12ms ± 1%  -14.49%  (p=0.000 n=10+9)
MVCCScan_RocksDB/rows=10000/versions=1/valueSize=64-8     2.66ms ± 2%    2.24ms ± 1%  -15.70%  (p=0.000 n=10+10)
MVCCScan_RocksDB/rows=10000/versions=1/valueSize=512-8    5.68ms ± 4%    3.69ms ± 1%  -35.00%  (p=0.000 n=10+9)
MVCCGet_RocksDB/versions=1/valueSize=8-8                  5.41µs ± 3%    5.44µs ± 2%     ~     (p=0.497 n=10+9)
MVCCGet_RocksDB/versions=10/valueSize=8-8                 6.21µs ± 2%    6.25µs ± 2%     ~     (p=0.252 n=9+10)
MVCCGet_RocksDB/versions=100/valueSize=8-8                17.4µs ±17%    18.5µs ±19%     ~     (p=0.218 n=10+10)
```

Release note (performance improvement): improve performance of scans by
copying less data in memory.

Co-authored-by: Jordan Lewis <jordanthelewis@gmail.com>
Co-authored-by: Peter Mattis <petermattis@gmail.com>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22267)
<!-- Reviewable:end -->

nvb committed:
storage: only access RangeStatsKey through StateLoader

Before this change, there were a number of places where we loaded and
stored MVCCStats directly from an engine. We now access them through
a StateLoader in all cases. This abstraction will allow us to minimize
the amount of logic needed when we migrate away from the use of this
key.

Release note: None

andy-kimball committed:
opt: Reorganize code to allow easier testing.

Before this change, the opt/xform tests were not able to use the
build.Builder. This is because the opt/build package is dependent
on opt/xform, and Golang doesn't allow circular dependencies.

This change reorganizes and updates files in order to break the
cycle. In particular, the Factory has been made into an interface
and placed in a new "opt" package, along with several other
foundational, but simple classes that many opt packages need to
share. The new set of directories looks like this:

build: build an opt expression from an AST
ops: contains the definitions for opt operators in Optgen format
opt: contains common interfaces and simple classes that other opt
     packages use to interface with one another
optgen: contains the Optgen language compiler and codegen tools
testutils: testing helper functions and packages
xform: the core optimizer files: factory, optimizer, and explorer

Note that the "opt" directory is temporary. Files in that
directory will be moved up one level once we've converted the
existing files (index constraints, old expr, etc.) to use the
new optimization methodology. See README.md in the "opt" directory
for more details.

Release note: None

Commit:
sql: add privilege checks on sequences

Before, they weren't checked -- nextval() and setval() just went right
through to the KV layer. Selecting from sequences wasn't checked either.

Release note (sql change): add privilege checks on sequences

petermattis committed:
libroach: use a fixed-length encoding for key/value lengths

```
Use a fixed-length encoding for key/value lengths returned for the
`MVCCScan` "batch" (note that this was already not the RocksDB batch
repr format).

name                                                    old time/op    new time/op    delta
MVCCScan_RocksDB/rows=100/versions=1/valueSize=8-8        28.9µs ± 1%    28.7µs ± 3%     ~     (p=0.353 n=10+10)
MVCCScan_RocksDB/rows=100/versions=1/valueSize=64-8       31.8µs ± 1%    30.9µs ± 2%   -2.93%  (p=0.000 n=10+10)
MVCCScan_RocksDB/rows=100/versions=1/valueSize=512-8      45.8µs ± 1%    44.2µs ± 1%   -3.41%  (p=0.000 n=10+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=8-8        221µs ± 0%     216µs ± 1%   -1.95%  (p=0.000 n=9+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=64-8       237µs ± 1%     230µs ± 1%   -2.98%  (p=0.000 n=10+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=512-8      371µs ± 1%     354µs ± 1%   -4.59%  (p=0.000 n=9+10)
MVCCScan_RocksDB/rows=10000/versions=1/valueSize=8-8      2.12ms ± 1%    2.07ms ± 1%   -2.42%  (p=0.000 n=9+10)
MVCCScan_RocksDB/rows=10000/versions=1/valueSize=64-8     2.24ms ± 1%    2.18ms ± 1%   -2.65%  (p=0.000 n=10+10)
MVCCScan_RocksDB/rows=10000/versions=1/valueSize=512-8    3.69ms ± 1%    3.35ms ± 2%   -9.35%  (p=0.000 n=9+10)
```

Release note: None

Commit:
distsql: fix TestFlowRegistryDrain flake

The flake was due to the flowRegistry returning from the drain process
before we had a chance to register an incoming flow. Added a hook that
is executed before the wait for any incoming flows and allowed this case
with a check for a minimum wait time.

Fixes #22240
Fixes #22239

Release note: None

andreimatei committed:
sql: teach the distSQLReceiver to discriminate between error types

Today our sql execution code doesn't discriminated between query
execution errors and errors in communication with the client (e.g. when
a query produces a result row and tries to send it to the client, at
which point it finds out that the client network connection is toast).
Communication errors are important because they mean that the sql
session is also toast - we should clean it up and stop processing
queries. So, communication errors, which originate in pgwire, then flow
all the way through sql execution back into pgwire, which recognizes
them and dismantles the session.
This scheme is already pretty funky today, and doesn't fit at all in the
future connExecutor code -> there a session is not directly driven by a
pgwire connection; the session runs its own goroutine communicating
async with pgwire. The connExecutor will be aware of the two different
types of errors and handle them differently.

In preparation for that, this patch makes code around the
distSQLReceiver able to discriminate between the two types of error - it
keeps track of communication errors versus all other errors.
The capability is not used yet in earnest, as both types of errors flow
to pgwire just like they did before, but the refactoring already
simplifies some code.

Release note: none

knz commented:
Made a mistake when creating this PR. Re-creating

nvb committed:
Merge pull request #22289 from nvanbenschoten/nvanbenschoten/accessRangeStats

storage: only access RangeStatsKey through StateLoader

vilterp committed:
Merge pull request #22284 from vilterp/seq-privilege-check

sql: add privilege checks on sequences

knz commented:
Not a mistake after all.

spencerkimball committed:
Merge pull request #21140 from spencerkimball/serializable-txn-improvements

Serializable txn improvements

couchand committed:
Merge pull request #22227 from couchand/chore/remove-dead-clusterviz-code

ui: remove dead clusterviz code

Commit:
Merge pull request #22237 from mrtracy/mtracy/ui_server_liveness

server/ui: Compute LivenessStatus on Server

Commit:
Merge pull request #22296 from asubiotto/flowdrainfix

distsql: fix TestFlowRegistryDrain flake

jordanlewis committed:
Merge pull request #22245 from jordanlewis/slicevector

storage/engine: use an append-only write batch

andreimatei committed:
Merge pull request #22306 from andreimatei/distsql-err

sql: teach the distSQLReceiver to discriminate between error types

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :ship: :it: 

---

Reviewed 53 of 53 files at r1.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22267#-:-L4IUMeFXduwidnRsXvl:b-tze86a)*
<!-- Sent from Reviewable.io -->

petermattis committed:
util/encoding: use stdlib binary.BigEndian.Uint{32,64} decoding

Use the stdlib `Uint{32,64}` decoding routines which generate better
code than what we previously had in
`DecodeUint{32,64}Ascending`. Presumably, the Go folks will keep that
code optimized.

name            old time/op  new time/op  delta
DecodeUint32-8  12.4ns ± 6%  12.3ns ± 4%     ~     (p=0.807 n=10+10)
DecodeUint64-8  13.8ns ± 2%  12.1ns ± 7%  -12.40%  (p=0.000 n=9+10)

Release note: none

knz committed:
sql: simplify valuesNode

the ValuesClause is not really needed after planning. Remove it.

Release note: none

knz committed:
sql: support arbitrary functions in FROM clauses

The following is valid in pg's SQL dialect:

```
> SELECT * FROM cos(123)
```

i.e. any scalar function is a valid data source.

This patch implements this in CockroachDB for compatibility.
This is needed e.g. by TypeORM.

Release note (sql change): scalar functions can now be used
in FROM clauses.

knz committed:
sql/parser: ensure that even functions that are keywords work in FROM

Release note: none

jordanlewis committed:
Merge pull request #22309 from jordanlewis/fixedlen

storage/engine: use a fixed-length encoding for key/value lengths

andy-kimball committed:
Merge pull request #22285 from andy-kimball/iface

opt: Reorganize code to allow easier testing.

petermattis committed:
Merge pull request #22313 from petermattis/pmattis/decode-uint64

util/encoding: use stdlib binary.BigEndian.Uint{32,64} decoding

knz committed:
Merge pull request #22314 from knz/20180201-values

sql: support arbitrary functions in FROM clauses

knz committed:
sql/sem/tree: make TableName harder to instantiate to non-default value

Suggested by @m-schneider; this will prevent usage mistakes.

Release note: None
